Dockerfile

```diff
@@ -16,28 +21,37 @@ ENV PYTHONUNBUFFERED=1 \
     POETRY_NO_INTERACTION=1 \
     VENV_PATH="/.venv"
 
-ENV PATH="$VENV_PATH/bin:$PATH"
+# ugly hack for C extension from lru-dict
+ENV CFLAGS="-g0 -O2 -ffile-prefix-map=/src=."
 
-FROM base as builder
+ENV PATH="$VENV_PATH/bin:$PATH"
 
 ENV POETRY_VERSION=1.3.2
-RUN pip install --no-cache-dir poetry==$POETRY_VERSION
+
+RUN pip install --no-compile --no-cache-dir poetry==$POETRY_VERSION
```

Comment:
I usually just remove poetry cache in the same stage (the same as with apt):

`RUN poetry install && rm ...`

Also, may be, you can get away with poetry from Debian:

`python3-poetry` package. There is some incompatibility between 1.8 and 2.0, I don't know what is used as base for python-3.12. Also, poetry (even old one from Debian packages) can be installed with `poetry self update`.